File: apps/minecraft/velocity/src/main/java/crabcraft/net/crabUtilities/velocity/NicknameCache.java
package crabcraft.net.crabUtilities.velocity;

import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;

public class NicknameCache {

    public record Snapshot(boolean loaded, String rawNickname, long version) {}

    private record Entry(boolean loaded, String rawNickname, long version) {}

    private final Map<UUID, Entry> nicknames = new ConcurrentHashMap<>();
    private final AtomicLong versions = new AtomicLong();

    public void setNickname(UUID uuid, String rawNickname) {
        String normalized = rawNickname == null ? "" : rawNickname;
        nicknames.compute(uuid, (id, current) ->
                new Entry(true, normalized, versions.incrementAndGet()));
    }

    public String getRawNickname(UUID uuid) {
        Entry entry = nicknames.get(uuid);
        return entry == null || !entry.loaded() || entry.rawNickname().isEmpty()
                ? null
                : entry.rawNickname();
    }

    public String getPlainNickname(UUID uuid) {
        String raw = getRawNickname(uuid);
        if (raw == null) return null;
        return stripColors(raw);
    }

    public boolean isLoaded(UUID uuid) {
        Entry entry = nicknames.get(uuid);
        return entry != null && entry.loaded();
    }

    public Snapshot snapshot(UUID uuid) {
        Entry entry = nicknames.get(uuid);
        return entry == null
                ? new Snapshot(false, null, 0L)
                : snapshot(entry);
    }

    public Snapshot beginLoad(UUID uuid) {
        Entry entry = nicknames.computeIfAbsent(uuid,
                id -> new Entry(false, "", versions.incrementAndGet()));
        return snapshot(entry);
    }

    public boolean commitIfVersion(UUID uuid, long expectedVersion, String rawNickname) {
        String normalized = rawNickname == null ? "" : rawNickname;
        AtomicBoolean committed = new AtomicBoolean();
        nicknames.compute(uuid, (id, current) -> {
            if (current == null || current.version() != expectedVersion) return current;
            committed.set(true);
            return new Entry(true, normalized, versions.incrementAndGet());
        });
        return committed.get();
    }

    public boolean discardIfUnloadedVersion(UUID uuid, long expectedVersion) {
        AtomicBoolean discarded = new AtomicBoolean();
        nicknames.computeIfPresent(uuid, (id, current) -> {
            if (current.loaded() || current.version() != expectedVersion) return current;
            discarded.set(true);
            return null;
        });
        return discarded.get();
    }

    public boolean isVersion(UUID uuid, long version) {
        Entry entry = nicknames.get(uuid);
        return entry != null && entry.loaded() && entry.version() == version;
    }

    public static String stripColors(String text) {
        return NicknameComponentParser.plain(text);
    }

    public void remove(UUID uuid) {
        nicknames.remove(uuid);
    }

    private static Snapshot snapshot(Entry entry) {
        return new Snapshot(entry.loaded(), entry.loaded() ? entry.rawNickname() : null, entry.version());
    }
}

File: apps/minecraft/velocity/src/main/kotlin/crabcraft/net/crabUtilities/velocity/NicknameCache.kt
package crabcraft.net.crabUtilities.velocity

import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong

open class NicknameCache {
    data class Snapshot(private val loaded: Boolean, private val rawNickname: String?, private val version: Long) {
        fun loaded() = loaded

        fun rawNickname() = rawNickname

        fun version() = version
    }

    private data class Entry(val loaded: Boolean, val rawNickname: String, val version: Long)

    private val nicknames = ConcurrentHashMap<UUID, Entry>()
    private val versions = AtomicLong()

    open fun setNickname(uuid: UUID, rawNickname: String?) {
        val normalised = rawNickname ?: ""
        nicknames.compute(uuid) { _, _ -> Entry(true, normalised, versions.incrementAndGet()) }
    }

    open fun getRawNickname(uuid: UUID): String? {
        val entry = nicknames[uuid]
        return if (entry == null || !entry.loaded || entry.rawNickname.isEmpty()) null else entry.rawNickname
    }

    open fun getPlainNickname(uuid: UUID): String? = getRawNickname(uuid)?.let(::stripColors)

    open fun isLoaded(uuid: UUID): Boolean = nicknames[uuid]?.loaded == true

    open fun snapshot(uuid: UUID): Snapshot = nicknames[uuid]?.let(::snapshot) ?: Snapshot(false, null, 0L)

    open fun beginLoad(uuid: UUID): Snapshot =
        snapshot(
            nicknames.computeIfAbsent(uuid) {
                Entry(false, "", versions.incrementAndGet())
            }
        )

    open fun commitIfVersion(uuid: UUID, expectedVersion: Long, rawNickname: String?): Boolean {
        val normalised = rawNickname ?: ""
        val committed = AtomicBoolean()
        nicknames.compute(uuid) { _, current ->
            if (current == null || current.version != expectedVersion) current
            else {
                committed.set(true)
                Entry(true, normalised, versions.incrementAndGet())
            }
        }
        return committed.get()
    }

    open fun discardIfUnloadedVersion(uuid: UUID, expectedVersion: Long): Boolean {
        val discarded = AtomicBoolean()
        nicknames.computeIfPresent(uuid) { _, current ->
            if (current.loaded || current.version != expectedVersion) current
            else {
                discarded.set(true)
                null
            }
        }
        return discarded.get()
    }

    open fun isVersion(uuid: UUID, version: Long): Boolean {
        val entry = nicknames[uuid]
        return entry != null && entry.loaded && entry.version == version
    }

    open fun remove(uuid: UUID) {
        nicknames.remove(uuid)
    }

    private fun snapshot(entry: Entry) =
        Snapshot(entry.loaded, entry.rawNickname.takeIf { entry.loaded }, entry.version)

    companion object {
        @JvmStatic fun stripColors(text: String?): String? = NicknameComponentParser.plain(text)
    }
}
